minimum_versions.py

```diff
@@ -0,0 +1,323 @@
+
+import rich_click as click
+import yaml
+from dateutil.relativedelta import relativedelta
+from rattler import Gateway, Version
+from rich.console import Console
+from rich.panel import Panel
+from rich.style import Style
+from rich.table import Column, Table
+from tlz.functoolz import curry, pipe
+from tlz.itertoolz import concat, groupby
+
+click.rich_click.SHOW_ARGUMENTS = True
+
+channels = ["conda-forge"]
+platforms = ["noarch", "linux-64"]
+ignored_packages = [
+    "coveralls",
+    "pip",
+    "pytest",
+    "pytest-cov",
+    "pytest-env",
+    "pytest-xdist",
+    "pytest-timeout",
+    "hypothesis",
+]
+
+
+@dataclass
+class Policy:
+    package_months: dict
+    default_months: int
+    overrides: dict[str, Version] = field(default_factory=dict)
+
+    def minimum_version(self, package_name, releases):
+        if (override := self.overrides.get(package_name)) is not None:
+            return override
+
+        policy_months = self.package_months.get(package_name, self.default_months)
+        today = datetime.date.today()
+
+        cutoff_date = today - relativedelta(months=policy_months)
+
+        index = bisect.bisect_left(
+            releases, cutoff_date, key=lambda x: x.timestamp.date()
+        )
+        return releases[index - 1 if index > 0 else 0]
+
+
+@dataclass
+class Spec:
+    name: str
+    version: Version | None
+
+    @classmethod
+    def parse(cls, spec_text):
+        warnings = []
+        if ">" in spec_text or "<" in spec_text:
+            warnings.append(
+                f"package should be pinned with an exact version: {spec_text!r}"
+            )
+
+            spec_text = spec_text.replace(">", "").replace("<", "")
+
+        if "=" in spec_text:
+            name, version_text = spec_text.split("=", maxsplit=1)
+            version = Version(version_text)
+            segments = version.segments()
+
+            if (len(segments) == 3 and segments[2] != [0]) or len(segments) > 3:
+                warnings.append(
+                    f"package should be pinned to a minor version (got {version})"
+                )
+        else:
+            name = spec_text
+            version = None
+
+        return cls(name, version), (name, warnings)
+
+
+@dataclass(order=True)
+class Release:
+    version: Version
+    build_number: int
+    timestamp: datetime.datetime = field(compare=False)
+
+    @classmethod
+    def from_repodata_record(cls, repo_data):
+        return cls(
+            version=repo_data.version,
+            build_number=repo_data.build_number,
+            timestamp=repo_data.timestamp,
+        )
+
+
+def parse_environment(text):
+    env = yaml.safe_load(text)
+
+    specs = []
+    warnings = []
+    for dep in env["dependencies"]:
+        spec, warnings_ = Spec.parse(dep)
+
+        warnings.append(warnings_)
+        specs.append(spec)
+
+    return specs, warnings
+
+
+def is_preview(version):
+    candidates = ["rc", "beta", "alpha"]
```

Comment:
`rattler` doesn't do anything different, the versions are those on `conda-forge`. I don't know if that always has the same rules as PyPI.

I haven't seen alpha/beta releases on `conda-forge`, yet, though, so those are there for completeness. I'd also go with a `set` literal instead of `tuple`/`list` literals.

Reply:
True.

```suggestion
    candidates = {"rc", "beta", "alpha"}
```

Reply:
I'll go with this, then:
```suggestion
    candidates = {"rc", "b", "a"}
```